### PR TITLE
[fix][broker] fix DEFAULT_NAR_EXTRACTION_DIR cause NoClassDefFoundError

### LINKED ISSUE #15978: [bug][broker] The DEFAULT_NAR_EXTRACTION_DIR will cause NoClassDefFoundError
**Describe the bug**
The  `DEFAULT_NAR_EXTRACTION_DIR = System.getProperty("java.io.tmpdir")` always use /tmp dir, os will auto cleanup unused file which is in  /tmp dir and unused in a few days. In this case, the protocol handler's .class file will be clean  and cause NoClassDefFoundError.

**Expected behavior**
The default  project tmp dir should be specified under the project home dir.

**Screenshots**
image

**Desktop (please complete the following information):**
 - OS: Centos7.2

### PR DESCRIPTION
**Pull request by Nicklee007**
Fixes #15978

### Motivation

There's an issue reported about problems in load protocol handler .class file. This PR will specify default tmp dir;

### Modifications
Specify default tmp dir by `-Djava.io.tmpdir=$PULSAR_HOME/tmp`

  
- [X] `doc-needed`

### CODE REVIEW

**Comment by gaozhangmin** [comment]
> DEFAULT_NAR_EXTRACTION_DIR

+1, we can config it through `narExtractionDirectory`

**Comment by Nicklee007** [comment]
> Would it make more sense to update the default for `DEFAULT_NAR_EXTRACTION_DIR` instead of changing the `java.io.tmpdir` location?

@michaeljmarshall @gaozhangmin  Update the `DEFAULT_NAR_EXTRACTION_DIR`  has smaller effect than change `java.io.tmpdir` ,I will change the default value of `DEFAULT_NAR_EXTRACTION_DIR`. Anyway default use the /tmp dir is not a good choice, the dir at risk be cleaned by os.

**Comment by Nicklee007** [comment]
@michaeljmarshall @gaozhangmin update `DEFAULT_NAR_EXTRACTION_DIR` default value ,which will choice `$PULSAR_HOME/tmp` dir by default, also user can specify the nar extraction dir by configure `narExtractionDirectory` in conf/broker.conf.  PTAL

File: pulsar-common/src/main/java/org/apache/pulsar/common/nar/NarClassLoader.java [hunk 1]
```diff
@@ -139,7 +139,7 @@ public class NarClassLoader extends URLClassLoader {
 
     private static final String TMP_DIR_PREFIX = "pulsar-nar";
 
-    public static final String DEFAULT_NAR_EXTRACTION_DIR = System.getProperty("java.io.tmpdir");
+    public static final String DEFAULT_NAR_EXTRACTION_DIR = System.getProperty("nar.extraction.tmpdir");
```
Comment (eolivelli) [inline pulsar-common/src/main/java/org/apache/pulsar/common/nar/NarClassLoader.java:142]: You have to fallback on the legacy name if this property is not configured

File: bin/pulsar [hunk 2]
```diff
@@ -290,6 +290,9 @@ if [[ -z "$IS_JAVA_8" ]]; then
   OPTS="$OPTS --add-opens java.base/sun.net=ALL-UNNAMED --add-opens java.management/sun.management=ALL-UNNAMED"
 fi
 
+# Specify default nar tmp dir
+OPTS="$OPTS -Dnar.extraction.tmpdir=$PULSAR_HOME/tmp"
```
Comment (eolivelli) [inline bin/pulsar:294]: Please don't change this default behaviour, we are going to break existing users that upgrade.

Now that you hBe the system property you can let your users configure it

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-common/src/main/java/org/apache/pulsar/common/nar/NarClassLoader.java (hunk 1 above)
Comment (Nicklee007) [reply pulsar-common/src/main/java/org/apache/pulsar/common/nar/NarClassLoader.java]: Has changed, if `nar.extraction.tmpdir` not configured will use the old config` java.io.tmpdir`.

File: bin/pulsar (hunk 2 above)
Comment (Nicklee007) [reply bin/pulsar:294]: Default use the /tmp dir has a risk which will cause the class file be auto cleaned by os, when the class file 
 used again , the class not find error occur . Now when the `nar.extraction.tmpdir` not configured will use the old config `java.io.tmpdir` , so the change will not break existing users again .

Comment (eolivelli) [reply bin/pulsar:294]: there is no guarantee that $PULSAR_HOME/tmp exists and it is writable by the user.
if we change the behaviour then we have to document it, but, as a general rule, we cannot break things during upgrades.
it is better to add this feature, in 2.11, then we can change the default behaviour in 2.12.

also, in linux /tmp is not emptied randomly, my experience it that such directory is empty at boot

**Review by eolivelli — CHANGES_REQUESTED** [review]
I support the change, as far as we don't change the behaviour at least for this release.

File: bin/pulsar (hunk 2 above)
Comment (Nicklee007) [reply bin/pulsar:294]: > it is better to add this feature, in 2.11, then we can change the default behaviour in 2.12.

Thx for the Suggestions.

> also, in linux /tmp is not emptied randomly, my experience it that such directory is empty at boot

maybe diff linux os has diff behaviours for the /tmp dir clean, Debian will clean when the boot , but as Centos or RedHat  will  auto clean the file in the /tmp dir when  7 days or 30 days the file not be used, In this way, the .class file will loss. 
In our case, we add the KOP protocol, after few days when we want use the KOP protocol, we find  some .class have been cleaned by os.

**Comment by Nicklee007** [comment]
> I support the change, as far as we don't change the behaviour at least for this release.

@eolivelli Could you help me add some label to mark the feature will add in the  follow release, Thx!

**Comment by eolivelli** [comment]
>  Could you help me add some label to mark the feature will add in the follow release, Thx!

There is actually now way to express that.
You can create a new issue to change the default value, and wait for 2.12 to start.

This is a OSS project, so it is hard to make plans for the future, as we are all volounteers.
I suggest you to create the issue and sponsor it when it will be the time.

I can't think of a better way currently

@merlimat do you have some better suggestion ? like creating 2.12 ahead of time ?

**Comment by Nicklee007** [comment]
> There is actually now way to express that.
> You can create a new issue to change the default value, and wait for 2.12 to start.

Thx for your suggestion, I have cancel to change the default nar tmp dir to `$PULSAR_HOME/tmp` in this PR. The relate PR #16143 will add in other release like 2.12.
Now, the  PR PTAL again, Thx.

**Review by eolivelli — APPROVED** [review]
LGTM

**Pull request merged**